extension/packagemanagerextension/serverextension/envmanager.py

```diff
@@ -0,0 +1,329 @@
+# pylint: disable=C0321
+
+import json
+import os
+import re
+import uuid
+import yaml
+
+from pkg_resources import parse_version
+from subprocess import check_output, CalledProcessError
+
+from traitlets.config.configurable import LoggingConfigurable
+from traitlets import Dict
+
+
+def pkg_info(s):
+    try:
+        # for conda >= 4.3, `s` should be a dict
+        name = s['name']
+        version = s['version']
+        build = s.get('build_string') or s['build']
+    except TypeError:
+        # parse legacy string version for information
+        name, version, build = s.rsplit('-', 2)
+
+    return {
+        'name': name,
+        'version': version,
+        'build': build
+    }
+
+
+def pkg_info_status(s, names):
+    try:
+        # for conda >= 4.3, `s` should be a dict
+        name = s['name']
+        version = s['version']
+        build = s.get('build_string') or s['build']
+    except TypeError:
+        # parse legacy string version for information
+        name, version, build = s.rsplit('-', 2)
+    status = "not available"
+    if name in names:
+        status = "installed"
+    return {
+        'name': name,
+        'version': version,
+        'build': build,
+        "status": status
+    }
+
+
+MAX_LOG_OUTPUT = 6000
+
+# try to match lines of json
+JSONISH_RE = r'(^\s*["\{\}\[\],\d])|(["\}\}\[\],\d]\s*$)'
+
+# these are the types of environments that can be created
+package_map = {
+    'python2': 'python=2 ipykernel',
+    'python3': 'python=3 ipykernel'
+}
+
+
+class EnvManager(LoggingConfigurable):
+    envs = Dict()
+
+    def _execute(self, cmd, *args):
+        cmdline = cmd.split() + list(args)
+        self.log.debug('[packagemanagerextension] command: %s', cmdline)
+
+        try:
+            output = check_output(cmdline)
+        except CalledProcessError as exc:
+            self.log.debug(
+                '[packagemanagerextension] exit code: %s', exc.returncode)
+            output = exc.output
+
+        self.log.debug('[packagemanagerextension] output: %s',
+                       output[:MAX_LOG_OUTPUT])
+
+        if len(output) > MAX_LOG_OUTPUT:
+            self.log.debug('[packagemanagerextension] ...')
+
+        return output.decode("utf-8")
+
+    def list_projects(self):
+        """List all projects that conda knows about"""
+        info = self.clean_conda_json(self._execute('conda info --json'))
+
+        def get_info(env):
+            return {
+                'name': os.path.basename(env),
+                'dir': env
+            }
+
+        return {
+            "projects": [get_info(env) for env in info['envs'] if "swanproject" in env]
+        }
+
+    def clean_conda_json(self, output):
```

Comment:
Do you know why there are cases we need to clean it? Does this come from the conda-jupy? Because, if it does, I would try to move to it some of these options (by importing) so that we can profit as much as possible from it and respective updates.